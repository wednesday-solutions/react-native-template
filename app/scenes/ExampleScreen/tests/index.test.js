import React from 'react';
import {
  RecoilRoot,
  useRecoilValueLoadable,
  useSetRecoilState,
  useRecoilState
} from 'recoil';
import { render, fireEvent } from '@testing-library/react-native';
import { ExampleScreenTest } from '../index';

jest.mock('recoil', () => ({
  ...jest.requireActual('recoil'),
  useRecoilValueLoadable: jest.fn(),
  useSetRecoilState: jest.fn(),
  useRecoilState: jest.fn()
}));
jest.mock('posthog-react-native', () => ({
  usePostHog: jest.fn(() => ({
    identify: jest.fn(),
    capture: jest.fn()
  }))
}));
describe('ExampleScreen', () => {
  const mockSetFetchTrigger = jest.fn();
  const mockSetUser = jest.fn();
  const mockUseRecoilState = jest.fn();

  beforeEach(() => {
    useSetRecoilState.mockReturnValue(mockSetFetchTrigger);
    useRecoilValueLoadable.mockReturnValue({ state: 'loading' });
    mockUseRecoilState.mockReturnValue([null, mockSetUser]);
    useRecoilState.mockImplementation(mockUseRecoilState);
    jest.spyOn(global.console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders loading indicator initially', () => {
    const { getByTestId } = render(
      <RecoilRoot>
        <ExampleScreenTest />
      </RecoilRoot>
    );

    expect(getByTestId('loader')).toBeTruthy();
  });

  test('renders content when userLoadable.state is "hasValue"', () => {
    useRecoilValueLoadable.mockReturnValue({
      state: 'hasValue',
      contents: { character: 'Homer' }
    });

    const { getByTestId } = render(
      <RecoilRoot>
        <ExampleScreenTest />
      </RecoilRoot>
    );

    expect(getByTestId('example-container-content')).toBeTruthy();
  });

  test('renders error message when userLoadable.state is "hasError"', () => {
    useRecoilValueLoadable.mockReturnValue({
      state: 'hasError',
      contents: { message: 'Error occurred' }
    });

    const { getByText } = render(
      <RecoilRoot>
        <ExampleScreenTest />
      </RecoilRoot>
    );

    expect(getByText('Error occurred')).toBeTruthy();
  });

  test('calls requestFetchUser on button press', () => {
    useRecoilValueLoadable.mockReturnValue({
      state: 'hasValue',
      contents: { character: 'Homer' }
    });

    const { getByText } = render(
      <RecoilRoot>
        <ExampleScreenTest />
      </RecoilRoot>
    );

    fireEvent.press(getByText('refresh'));
    expect(mockSetFetchTrigger).toHaveBeenCalled();
  });

  test('sets user when userLoadable.state is "hasValue"', () => {
    useRecoilValueLoadable.mockReturnValue({
      state: 'hasValue',
      contents: { character: 'Homer' }
    });

    render(
      <RecoilRoot>
        <ExampleScreenTest />
      </RecoilRoot>
    );

    expect(mockSetUser).toHaveBeenCalledWith({ character: 'Homer' });
  });
});
