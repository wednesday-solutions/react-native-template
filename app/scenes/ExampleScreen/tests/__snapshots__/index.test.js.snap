// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExampleScreen /> Container tests should render and match the snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "marginBottom": 30,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 30,
        "maxWidth": 320,
      },
    ]
  }
>
  <View
    testID="example-container-content"
  >
    <View>
      <View
        style={
          Object {
            "height": 150,
            "marginBottom": 25,
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../app/assets/images/wednesday-logo.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 17,
              "fontStyle": "italic",
              "fontWeight": "normal",
              "marginBottom": 5,
              "textAlign": "center",
            },
          ]
        }
      >
        Press Cmd+R to reload,
Cmd+D or shake for dev menu.
      </Text>
    </View>
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 17,
              "fontWeight": "normal",
              "marginBottom": 5,
              "textAlign": "center",
            },
          ]
        }
      >
        character loves Wednesday
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontSize": 17,
              "fontWeight": "normal",
              "marginBottom": 5,
              "textAlign": "center",
            },
          ]
        }
      >
        because
      </Text>
      <Image
        resizeMode="contain"
        source={
          Object {
            "uri": undefined,
          }
        }
        style={
          Array [
            Object {
              "height": 80,
              "marginBottom": 0,
              "marginLeft": "auto",
              "marginRight": "auto",
              "marginTop": 0,
              "width": 80,
            },
          ]
        }
        testID="character-image"
      />
      <Text
        style={
          Array [
            Object {
              "fontSize": 17,
              "fontWeight": "normal",
              "marginBottom": 5,
              "textAlign": "center",
            },
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "alignSelf": "center",
            "marginTop": 40,
            "maxWidth": 80,
          },
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Refresh
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
